CMakeLists.txt

```diff
@@ -217,44 +219,72 @@ message(STATUS "Setting link libraries")
 
 target_link_libraries(Hyprland rt PkgConfig::deps)
 
+# list of headers
+set(PROTOCOL_HEADERS "" CACHE INTERNAL "List of protocol headers")
+
 function(protocol protoPath protoName external)
     if (external)
-        execute_process(
-            COMMAND ${WaylandScanner} server-header ${protoPath} protocols/${protoName}-protocol.h
-            WORKING_DIRECTORY ${CMAKE_SOURCE_DIR})
-        execute_process(
-            COMMAND ${WaylandScanner} private-code ${protoPath} protocols/${protoName}-protocol.c
-            WORKING_DIRECTORY ${CMAKE_SOURCE_DIR})  
-        target_sources(Hyprland PRIVATE protocols/${protoName}-protocol.c)
+        set(path ${CMAKE_SOURCE_DIR}/${protoPath})
     else()
-        execute_process(
-            COMMAND ${WaylandScanner} server-header ${WAYLAND_PROTOCOLS_DIR}/${protoPath} protocols/${protoName}-protocol.h
-            WORKING_DIRECTORY ${CMAKE_SOURCE_DIR})
-        execute_process(
-            COMMAND ${WaylandScanner} private-code ${WAYLAND_PROTOCOLS_DIR}/${protoPath} protocols/${protoName}-protocol.c
-            WORKING_DIRECTORY ${CMAKE_SOURCE_DIR})  
-        target_sources(Hyprland PRIVATE protocols/${protoName}-protocol.c)
+        set(path ${WAYLAND_PROTOCOLS_DIR}/${protoPath})
     endif()
+
+    add_custom_command(
+        OUTPUT ${CMAKE_SOURCE_DIR}/protocols/${protoName}-protocol.h
+        COMMAND ${WaylandScanner} server-header ${path} protocols/${protoName}-protocol.h
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+    )
+    add_custom_command(
+        OUTPUT ${CMAKE_SOURCE_DIR}/protocols/${protoName}-protocol.c
+        COMMAND ${WaylandScanner} private-code ${path} protocols/${protoName}-protocol.c
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+    )
+    target_sources(Hyprland PRIVATE ${CMAKE_SOURCE_DIR}/protocols/${protoName}-protocol.h)
+    target_sources(Hyprland PRIVATE ${CMAKE_SOURCE_DIR}/protocols/${protoName}-protocol.c)
```

Comment:
you can call `target_sources(generate-protocols-headers PRIVATE ...)` here, so no need to use `PROTOCOL_HEADERS` variable. This should be a bit cleaner IMO.

Also, `target_sources` accepts multiple files :)